sslh-select.c

```diff
@@ -180,6 +182,93 @@ void shovel(struct connection *cnx, int active_fd,
     }
 }
 
+/* shovels data from one fd to the other and vice-versa
+   returns after one socket closed
+ */
+void shovel_single(struct connection *cnx)
+{
+   fd_set fds_r, fds_w;
+   int res, i;
+   int max_fd = MAX(cnx->q[0].fd, cnx->q[1].fd) + 1;
+
+   FD_ZERO(&fds_r);
+   FD_ZERO(&fds_w);
+   while (1) {
+      for (i = 0; i < 2; i++) {
+         if (cnx->q[i].deferred_data_size) {
+            FD_SET(cnx->q[i].fd, &fds_w);
+            FD_CLR(cnx->q[1-i].fd, &fds_r);
+         } else {
+            FD_CLR(cnx->q[i].fd, &fds_w);
+            FD_SET(cnx->q[1-i].fd, &fds_r);
+         }
+      }
+
+      res = select(
+                   max_fd,
+                   &fds_r,
+                   &fds_w,
+                   NULL,
+                   NULL
+                  );
+      CHECK_RES_DIE(res, "select");
+
+      for (i = 0; i < 2; i++) {
+          if (FD_ISSET(cnx->q[i].fd, &fds_w)) {
+              res = flush_deferred(&cnx->q[i]);
+              if ((res == -1) && ((errno == EPIPE) || (errno == ECONNRESET))) {
+                  if (verbose)
+                      fprintf(stderr, "%s socket closed\n", i ? "server" : "client");
+                  return;
+              }
+          }
+          if (FD_ISSET(cnx->q[i].fd, &fds_r)) {
+              res = fd2fd(&cnx->q[1-i], &cnx->q[i]);
+              if (!res) {
+                  if (verbose)
+                      fprintf(stderr, "socket closed\n");
+                  return;
+              }
+          }
+      }
+   }
+}
+
+/* Child process that makes internal connection and proxies
+ */
+void connect_proxy(struct connection *cnx)
+{
+    int in_socket;
+    int out_socket;
+
+    /* Minimize the file descriptor value to help select() */
+    in_socket = dup(cnx->q[0].fd);
```

Comment:
Nice :-) naive question: does that noticebly increase performance?

Reply:
Honestly: I don’t know :-) But I remember reading again and again that `select` is slow because it has to go through all of the file descriptors in the set and that’s one of the reasons why other notification mechanisms like kqueue and epoll were introduced. I’m not sure whether it depends on the _total_ size of the set or only on the input number of _active_ descriptors though.

Reply:
Depends what you mean by "total size": a fd_set contains a fixed number of file descriptors (FD_SETSIZE, 1024 in Linux), but select only polls up to those marked useful in the first argument, so:
select(max, fds1, fds2, fds3, ts);
for(i = 0; i < max; i++) {
   if (fds1 & (1 << i)) {
   }
}

So the problem is mostly if you have a lot of descriptors but most are never active, so select() wakes up and loops over a thousand fds just to find one active bit.